ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,605 @@
+)
+worker.run(shutdown_signals: ['SIGINT'])
+```
+
+Example client usage:
+
+```ruby
+client = Temporalio::Client.connect('localhost:7233', 'my-namespace')
+
+# Start workflow
+wf_handle = client.start_workflow(
+  GreetingWorkflow,
+  id: 'my-workflow-id',
+  task_queue: 'my-task-queue'
+)
+
+# Send query
+supported_languages = wf_handle.query(
+  GreetingWorkflow.languages,
+  { include_unsupported: false }
+)
+puts "Supported languages: #{supported_languages}"
+
+# Execute update
+previous_language = wf_handle.execute_update(
+  GreetingWorkflow.set_language,
+  :chinese
+)
+current_language = wf_handle.query(GreetingWorkflow.language)
+puts "Language changed: #{previous_language} -> #{current_language}"
+
+# Start update then wait for complete
+update_handle = wf_handle.start_update(
+  GreetingWorkflow.set_language_using_activity,
+  :arabic,
+  wait_for_stage: Temporalio::Client::WorkflowUpdateWaitStage::ACCEPTED
+)
+previous_language = update_handle.result
+current_language = wf_handle.query(GreetingWorkflow.language)
+puts "Language changed: #{previous_language} -> #{current_language}"
+
+# Send signal
+wf_handle.signal(
+  GreetingWorkflow.approve,
+  { name: 'some-approver-name' }
+)
+puts "Workflow result: #{wf_handle.result}"
+```
+
+Things to note about workflow definition:
+
+* Workflows must extend (directly or indirectly) from `Temporalio::Workflow::Definition`.
+  * It must implement `execute` which is the entry point.
+  * Workflow can customize its name via `workflow_name <name>` in the workflow somewhere.
+  * **NOTE** As part of this proposal, we are changing the activity base class from `Temporalio::Activity` class to
+    `Temporalio::Activity::Definition` to be consistent with the base class for workflows.
+  * 💭 Why a class?
+    * This fits well with the Ruby mindset and many of our SDKs utilize classes to encapsulate workflows and their
+      handlers/state.
+* Workflow has several class methods that start with `workflow_` that can only be called during class definition time
+  not runtime. See next section for list.
+* When used in a client:
+  * The class of the workflow can be provided instead of the name, e.g. `my_client.start_workflow(MyWorkflowClass, ...)`
+  * The signal, query, or update can reference a no-arg class method of the same name as the instance method, e.g.
+    `my_wf_handle.signal(MyWorkflowClass.my_signal, ...)`, which returns a signal/query/update definition
+    * Are we ok with `respond_to_missing?` and `inherited` for class method access of these values? Is this too much
+      magic?
+* Workflow cannot be instantiated by users (i.e. outside of workflow environment)
+
+## Workflow API
+
+Approximate guess of how the `Workflow` module may look:
+
+```ruby
+module Temporalio
+  module Workflow
+
+    # Class users must extend for a workflow
+    class Definition
+      class << self
+        # Protected because they are only for this implementation
+        protected
```

Comment:
nit: You probably mean `private` as I don't think you need one instance accessing another instance's methods.
(`protected` is very niche in ruby, and `private` means what `protected` means in .NET)

Reply:
I do mean `protected` very intentionally and am aware of its quirks and its rareness in Ruby use (we don't use it elsewhere). These are for use by a workflow, e.g.

```ruby
class MyWorkflow < Temporalio::Workflow::Definition
  workflow_name 'my-custom-workflow-name'

  # ...
end
```

If `workflow_name` were private, this would fail. If `workflow_name` were public, then it will be easily invocable from the client side, but we only want signals, updates, and queries exposed as class methods client side (that return definitions), e.g. `my_handle.signal(MyWorkflow.my_signal, 'my-signal-arg')`. So in this case, protected makes sense. We considered putting these in a mixin instead, but it was unnecessary since we also want class extension anyways.